********************
Command Line Options
********************


.. index:: Input/Output Options

Input/Output and Execution Options
**********************************

List of all arguments related to input/output and general execution
of the program.

- ``-h``, ``--help``

  Show help messages for VaxPress and exit.

--------------------
Input/Output Options
--------------------

==================
Required Arguments
==================

- ``-i FILE``, ``--input FILE``

  Path to the input FASTA file containing the CDS or protein sequence.

- ``-o DIR``, ``--output DIR``

  Path to the output directory.

==================
Optional Arguments
==================

- ``--protein``

  Input is a protein sequence.

- ``--overwrite``

  Overwrite output directory if it already exists.

- ``-q``, ``--quiet``

  Do not print progress messages to stdout.

- ``--print-top N``

  Print evaluation results of top ``N`` sequences at the ends of
  iterations (default: 10).

- ``--report-interval MIN``

  Interval of report updates in minutes. In other words, ``report.html``
  will be updated by this time interval (default: 5).

- ``--version``

  Show program's version number and exit.

.. index:: Execution Options
.. _execution options:

-----------------
Execution Options
-----------------

- ``-p N``, ``--processes N``

  Number of processes to use (default: 4).

- ``--preset FILE``
  
  Use pre-defined options stored in ``FILE``, which has the same
  format and content as the ``parameters.json`` file generated by
  a VaxPress run.  Any options specified in the command line will
  supersede these preset values.

  Example command to use a preset file::

    vaxpress -i {path_to_input.fa} \
             -o {path_to_output_directory} \
             --preset {path_to_parameters.json}

  Refer to :ref:`label-preset` for more information.

- ``--addon FILE``

  Load a custom fitness function.

  Example command to use a custom fitness function::

    vaxpress -i {path_to_input.fa} \
             -o {path_to_output_directory} \
             --iterations {n_iterations} \
             -p {n_processes} \
             --addon {path_to_addon.py}

  Please refer to the :ref:`label-addon` section of this documentation
  for detailed instructions on using a custom fitness function.

- ``--seed NUMBER``

  Alter the random seed ``NUMBER`` to generate a different sequence
  under the same settings (default: 922).

- ``--folding-engine NAME``

  RNA folding engine: ``vienna`` or ``linearfold`` (default: ``vienna``).
  ``linearfold`` is only available when a separate Python package,
  `linearfold-unofficial <https://pypi.org/project/linearfold-unofficial/>`_,
  is installed.

- ``--default-off``

  Disable all fitness functions by default. This is useful
  when you wish to activate only a select few fitness functions.

.. index:: Optimization Options


Optimization Options
********************

The following is a list of parameters related to the program's
optimization processes. Refer to the :ref:`tuning-parameters` section
for examples demonstrating the impact of each parameter on the
optimization process.

- ``--random-initialization``

  Randomize all codons at the beginning (default: False).

.. _label-constart:

- ``--conservative-start ITER[:WIDTH]``
  
  Conserve sequence for the first ``ITER`` iterations, except the first
  ``WIDTH`` amino acids (default width: 7). It's a recommended option
  to use when running VaxPress with LinearDesign initialization.
  See :ref:`using-lineardesign` for more information.

  .. code-block:: bash

    # Example command line to use --conservative-start option
    vaxpress -i ./testseq/spike.fa \
             -o ../test_run \
             --iterations 1000 \
             --lineardesign 1 \
             --lineardesign-dir ../LinearDesign \
             --conservative-start 10:7 \

- ``--iterations N``

  Number of iterations (default: 10).

- ``--population N``

  Population size to generate every iteration (default: 20).

- ``--survivors N``

  Number of survivors to pass to the next iteration (default: 2).

- ``--initial-mutation-rate RATE``

  Initial mutation rate (default: 0.1).

- ``--full-scan-interval N``
  
  Number of iterations between full scans of single mutations of
  unpaired bases (default: 300).

- ``--winddown-trigger N``

  Number of iterations with the same best score to trigger mutation
  stabilization (default: 15). Please refer to
  :ref:`label_WinddownTR` for detailed explanation.

- ``--winddown-rate RATE``

  Mutation rate multiplier when mutation stabilization is triggered
  (default: 0.9). Please refer to
  :ref:`label_WinddownTR` for detailed explanation.

- ``--boost-loop-mutations WEIGHT[:START]``

  Boost mutations in loops after position ``START`` by ``WEIGHT``
  (default: 1.5:15).

- ``--species NAME``

  Target species (default: human).

- ``--codon-table NAME``

  Codon table that VaxPress refers to (default: standard). Codon
  tables are imported from `Bio.Data.CodonTable module
  <https://biopython.org/docs/1.75/api/Bio.Data.CodonTable.html>`_. To
  check for the full list of supported codon tables, please refer
  to `BioPython source code
  <https://github.com/biopython/biopython/blob/master/Bio/Data/CodonTable.py>`_.

.. index:: LinearDesign; Options
.. _label-linopts:

LinearDesign Options 
********************

- ``--lineardesign LAMBDA``

  Call LinearDesign to initialize the optimization. ``LAMBDA`` (λ)
  is a parameter specifying the ratio that MFE and CAI are reflected
  in the optimization. λ is in (–∞, 0] while λ = 0 means only MFE
  is considered, and the weight on CAI increases as λ increases.

- ``--lineardesign-dir DIR``

  Path to the top directory containing LinearDesign. When this
  argument is specified, its information will be stored in
  ``config.json`` :ref:`file <label-configuration>` inside VaxPress
  program. As a result, you don't need to write this option again
  from the next time if you're using the same LinearDesign directory.

- ``--lineardesign-omit-start AA``

  The number of amino acids to omit from the N-terminus when calling
  LinearDesign (default: 5). By using this option, generation of
  folded structures in start codon region while optimizing MFE by
  LinearDesign can be avoided.

.. index:: Fitness Function Options

Options Related to Fitness Functions
************************************

List of all arguments related to fitness functions inside VaxPress.
For detailed information on how each scoring function works, see
:doc:`Agorithmic Details </algorithmic_details>`.

------
iCodon
------

- ``--iCodon-weight WEIGHT``
  
  Scoring weight for iCodon predicted stability (default: 1.0).

--------
DegScore 
--------

- ``--degscore-weight WEIGHT``
  
  Set DegScore scoring weight (default: 0.0). Setting this weight
  as certain positive value, VaxPress optimizes sequence toward the
  direction that minimize DegScore value. If you want to make
  DegScore value influential, it's recommended to set the weight
  at least 5.0.

----------------------
Codon Adaptation Index
----------------------

- ``--cai-weight WEIGHT``
  
  Scoring weight for codon adaptation index (default: 3.0).

-------------------------------------
Codon Adaptation Index of Codon-Pairs
-------------------------------------

- ``--bicodon-weight WEIGHT``
  
  Scoring weight for codon adaptation index of codon-pairs (default: 1.0).

--------
Uridines
--------

- ``--ucount-weight WEIGHT``
  
  Scoring weight for U count minimizer (default: 3.0).

-----------
RNA Folding
-----------

===
MFE
===

- ``--mfe-weight WEIGHT``
  
  Scoring weight for Minimum Free Energy (MFE) (default: 3.0).


=====
Loops
=====

- ``--loop-weight WEIGHT``
  
  Scoring weight for loops (default: 1.5).

- ``--loop-threshold N``
  
  Minimum count of unfolded bases to be considered as a loop (default: 2).

==========================
Structure near Start Codon
==========================

- ``--start-str-weight WEIGHT``
  
  Penalty weight for folded start codon region (default: 1).

- ``--start-str-width WIDTH``
  
  Width in nt of unfolded region near the start codon (default: 15).

==========
Long Stems
==========

- ``--longstem-weight WEIGHT``
  
  Penalty score for long stems (default: 100.0).

- ``--longstem-threshold N``
  
  Minimum length of stems to avoid (default: 27).

--------------
Local GC Ratio
--------------

- ``--gc-weight WEIGHT``
  
  Scoring weight for GC ratio (default: 3.0).

- ``--gc-window-size SIZE``
  
  Size of window for GC content calculation (default: 50).

- ``--gc-stride STRIDE``
  
  Size of stride for GC content calculation (default: 5).

--------------
Tandem Repeats
--------------

- ``--repeats-weight WEIGHT``
  
  Scoring weight for tandem repeats (default: 1.0).

- ``--repeats-min-repeats N``
  
  Minimum number of repeats to be considered as a tandem repeat (default: 2).

- ``--repeats-min-length LENGTH``
  
  Minimum length of repeats to be considered as a tandem repeat (default: 10).